src/mawpy/steps/trace_segmentation_clustering.py

```diff
@@ -21,9 +21,15 @@ def _get_diameter_constraint_exceed_index(starting_index: int, point_to_check: i
 
         If no such points are found between [starting_index, point_to_check) then return -1
     """
+
+    # Map to store distance values to avoid re-computation for duplicate points
+    distance_map = {}
     for i in range(starting_index, point_to_check):
-        if spatial_constraint < distance(latitudes_list[i], longitudes_list[i],
-                                         latitudes_list[point_to_check], longitudes_list[point_to_check]):
+        point_key = (latitudes_list[i], longitudes_list[i])
+        if point_key not in distance_map:
+            distance_map[point_key] = distance(latitudes_list[i], longitudes_list[i],
+                                               latitudes_list[point_to_check], longitudes_list[point_to_check])
+        if spatial_constraint < distance_map[point_key]:
```

Comment:
This is likely very minor, but you can avoid doing the map lookup twice
```suggestion
        d = distance_map.get(point_key, -1)
        if d == -1:
            d = distance(latitudes_list[i], longitudes_list[i],
                                               latitudes_list[point_to_check], longitudes_list[point_to_check])
            distance_map[point_key] = d
        if spatial_constraint < d:
```